x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.LongObjectPagedHashMap;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorBase;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.xcontent.ToXContent;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.REGISTRY_KEY;
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.SUPPORTED_TYPES;
+
+public abstract class MapReduceAggregator<
+    MapContext extends Closeable,
+    MapFinalContext extends Writeable,
+    ReduceContext extends Closeable,
+    Result extends ToXContent & Writeable> extends AggregatorBase {
+
+    private final List<ValuesExtractor> extractors;
+    private final List<String> fieldNames;
+    private final AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer;
+    private final BigArrays bigArraysForMapReduce;
+    private final LongObjectPagedHashMap<Object> mapReduceContextByBucketOrdinal;
+    private final boolean profiling;
+    private final DelegatingCircuitBreakerService breakerService;
+
+    protected MapReduceAggregator(
+        String name,
+        AggregationContext context,
+        Aggregator parent,
+        Map<String, Object> metadata,
+        AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer,
+        List<ValuesSourceConfig> configs
+    ) throws IOException {
+        super(name, AggregatorFactories.EMPTY, context, parent, CardinalityUpperBound.NONE, metadata);
+
+        List<ValuesExtractor> extractors = new ArrayList<>();
+        List<String> fieldNames = new ArrayList<>();
+        int id = 0;
+        for (ValuesSourceConfig c : configs) {
+            if (SUPPORTED_TYPES.contains(c.valueSourceType()) == false) {
+                throw new IllegalArgumentException(
+                    c.getDescription() + " is not supported for aggregation [" + mapReducer.getWriteableName() + "]"
+                );
+            }
+
+            ValuesExtractor e = context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, c).build(c, id++);
```

Comment:
It's weird to me that we do this here instead of in the aggregator factory.  We're adding this map reduce layer because we want future flexibility to extend the algorithm selection, but here we're saying that all future algorithms will use the same values extractors.  I don't see how we can know that.  It makes more sense to me for the factory, which is generally responsible for algorithm selection, to get the appropriate ValueExtractors.

Another way to do that would be setting the REGISTRY_KEY based on the name of the algorithm.  Then when we pull the extractors out here, we get the ones specifically for this algorithm.

Reply:
I've moved this into the agg.